frontendUiTable/src/main/java/com/virtuslab/gitmachete/frontend/ui/VcsRootComboBox.java

```diff
@@ -0,0 +1,90 @@
+package com.virtuslab.gitmachete.frontend.ui;
+
+import java.awt.BorderLayout;
+import java.util.List;
+
+import javax.swing.Box;
+import javax.swing.JComboBox;
+import javax.swing.JComponent;
+import javax.swing.JPanel;
+
+import com.intellij.ui.MutableCollectionComboBoxModel;
+import com.intellij.ui.SimpleListCellRenderer;
+import com.intellij.util.SmartList;
+import git4idea.repo.GitRepository;
+import org.checkerframework.checker.guieffect.qual.UIEffect;
+import org.checkerframework.common.value.qual.MinLen;
+
+import com.virtuslab.gitmachete.frontend.ui.selection.ISelectionChangeObservable;
+import com.virtuslab.gitmachete.frontend.ui.selection.ISelectionChangeObserver;
+
+public final class VcsRootComboBox extends JComboBox<GitRepository>
+    implements
+      ISelectionChangeObservable<GitRepository> {
+  private final List<ISelectionChangeObserver> observers = new SmartList<>();
+
+  /**
+   * @param repositories non-empty list of {@link git4idea.repo.GitRepository} that represents VCS repositories
+   */
+  @UIEffect
+  public VcsRootComboBox(@MinLen(1) List<GitRepository> repositories) {
+    super(new MutableCollectionComboBoxModel<>(/* items */ repositories, /* selection */ repositories.get(0)));
+    setRenderer(SimpleListCellRenderer.create("", repo -> repo.getRoot().getName()));
+  }
+
+  @Override
+  public MutableCollectionComboBoxModel<GitRepository> getModel() {
+    return (MutableCollectionComboBoxModel<GitRepository>) super.getModel();
```

Comment:
Hmm maybe to avoid downcast I'd extract the `model` to a private final field

Reply:
This would cause a situation where there is no appropriate model set (before `setModel` in the constructor). Furthermore, it seems to cause some strange `java.lang.IllegalArgumentException: dataModel must be non null`. I would prefer to keep the current solution.

Reply:
Apparently the runtime non-null check is generated by some JetBrains' annotation processor for `@org.jetbrains.annotations.NotNull`-marked params & results... which is kinda necessary since they don't have any means of consistently enforcing that non-nullness on the library consumers in the compile time.

Reply:
(at least wrt. params; non-null return values can be ofc always enforced in the compile time)